distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/MachineDependentHeap.java

```diff
@@ -8,193 +8,153 @@
-         * would cause a discrepancy for people using that version of Java. But there's no harm
-         * in a bit of extra rounding here - it can only reduce discrepancies.
-         *
-         * If this formula is changed then corresponding changes must be made to the {@code NativeMemoryCalculator} and
-         * {@code MlAutoscalingDeciderServiceTests} classes in the ML plugin code. Failure to keep the logic synchronized
-         * could result in repeated autoscaling up and down.
-         */
-        ML_ONLY(m -> mb(m <= (GB * 16) ? (long) (m * .4) : (long) min((GB * 16) * .4 + (m - GB * 16) * .1, MAX_HEAP_SIZE), 4)),
-
-        /**
-         * Data node. Essentially any node that isn't a master or ML only node.
-         *
-         * <p>Heap is computed as:
-         * <ul>
-         *     <li>40% of total system memory when less than 1 gigabyte with a minimum of 128 megabytes.</li>
-         *     <li>50% of total system memory when greater than 1 gigabyte up to a maximum of 31 gigabytes.</li>
-         * </ul>
-         */
-        DATA(m -> mb(m < GB ? max((long) (m * .4), MIN_HEAP_SIZE) : min((long) (m * .5), MAX_HEAP_SIZE))),
-
-        /**
-         * Unknown role node.
-         *
-         * <p>Hard-code heap to a default of 1 gigabyte.
-         */
-        UNKNOWN(m -> DEFAULT_HEAP_SIZE_MB);
+    protected static int mb(long bytes) {
+        return (int) (bytes / (1024 * 1024));
+    }
 
-        private final Function<Long, Integer> formula;
+    protected static int mb(long bytes, int toLowerMultipleOfMb) {
+        return toLowerMultipleOfMb * (int) (bytes / (1024 * 1024 * toLowerMultipleOfMb));
+    }
 
-        MachineNodeRole(Function<Long, Integer> formula) {
-            this.formula = formula;
+    private static MachineNodeRole mapNodeRole(List<DiscoveryNodeRole> roles) {
+        if (roles.isEmpty()) {
+            // If roles are missing or empty (coordinating node) assume defaults and consider this a data node
+            return MachineNodeRole.DATA;
+        } else if (containsOnly(roles, MASTER_ROLE)) {
+            return MachineNodeRole.MASTER_ONLY;
+        } else if (roles.contains(ML_ROLE) && containsOnly(roles, ML_ROLE, REMOTE_CLUSTER_CLIENT_ROLE)) {
+            return MachineNodeRole.ML_ONLY;
+        } else {
+            return MachineNodeRole.DATA;
         }
+    }
 
-        /**
-         * Determine the appropriate heap size for the given role and available system memory.
-         *
-         * @param systemMemory total available system memory in bytes
-         * @return recommended heap size in megabytes
-         */
-        public int heap(long systemMemory) {
-            return formula.apply(systemMemory);
-        }
+    @SuppressWarnings("unchecked")
+    private static <T> boolean containsOnly(Collection<T> collection, T... items) {
+        return Arrays.asList(items).containsAll(collection);
```

Comment:
nit - Maybe `items` could be passed in as a `Set` so the `containsAll` isn't doing a list vs list? May not be worth the additional effort though